finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+        }
+        let node_id = authentication.0.creator();
+        let rebroadcast = match self.last_broadcast.get(&node_id) {
+            Some(instant) => time::Instant::now() > *instant + self.cooldown,
+            None => true,
+        };
+        let message = match rebroadcast {
+            true => {
+                self.last_broadcast.insert(node_id, time::Instant::now());
+                Some(authentication_broadcast(authentication))
+            }
+            false => None,
+        };
+        (addresses, message)
+    }
+
+    fn create_response(
+        &mut self,
+        requester_id: NodeIndex,
+        node_ids: Vec<NodeIndex>,
+        handler: &mut SessionHandler,
+    ) -> Option<(DiscoveryMessage, DataCommand)> {
+        let requested_authentications = self
+            .requested_authentications
+            .entry(requester_id)
+            .or_default();
+        requested_authentications.extend(node_ids);
+        if let Some(instant) = self.last_response.get(&requester_id) {
+            if time::Instant::now() < *instant + self.cooldown {
+                return None;
+            }
+        }
+        let peer_id = match handler.peer_id(&requester_id) {
+            Some(peer_id) => peer_id,
+            None => return None,
+        };
+        let authentications: Vec<_> = requested_authentications
+            .iter()
+            .filter_map(|id| handler.authentication_for(id))
+            .collect();
+        if authentications.is_empty() {
+            None
+        } else {
+            self.last_response
+                .insert(requester_id, time::Instant::now());
+            self.requested_authentications.remove(&requester_id);
+            Some(response(authentications, peer_id))
+        }
+    }
+
+    fn handle_request(
+        &mut self,
+        node_ids: Vec<NodeIndex>,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> (Vec<Multiaddr>, Option<(DiscoveryMessage, DataCommand)>) {
+        let node_id = authentication.0.creator();
+        let addresses = self.handle_authentication(authentication, handler);
+        if addresses.is_empty() {
+            return (addresses, None);
+        }
+        (addresses, self.create_response(node_id, node_ids, handler))
+    }
+
+    fn handle_response(
+        &mut self,
+        authentications: Vec<Authentication>,
+        handler: &mut SessionHandler,
+    ) -> Vec<Multiaddr> {
+        authentications
+            .into_iter()
+            .flat_map(|authentication| self.handle_authentication(authentication, handler))
+            .collect()
+    }
+
+    /// Analyzes the provided message and returns all the new multiaddresses we should
+    /// be connected to and any messages that we should send as a result of it.
+    pub fn handle_message(
+        &mut self,
+        message: DiscoveryMessage,
+        handler: &mut SessionHandler,
+    ) -> (Vec<Multiaddr>, Option<(DiscoveryMessage, DataCommand)>) {
```

Comment:
I woudl go for an alias for `(Vec<Multiaddr>, Option<(DiscoveryMessage, DataCommand)>)` - it would both better document the auxiliary functions and make the public API more descriptive

Reply:
I think after the change you proposed above, making an alias for the type inside the internal parentheses, aliasing this would obfuscate more than help.